### PR TITLE
Added Missing Regex for Allowlist Dependency Files

### PR DESCRIPTION
**Pull request by honeyankit**
### What are you trying to accomplish?

This pull request introduces improvements to the `updated_files_regex` method across all file updaters, enhancing their file matching capabilities. The `updated_files_regex` method within the FileUpdater class will now ensure that only the intended files are updated for each specific ecosystem. Additionally, this PR includes test cases for the `updated_files_regex` method across all 18 ecosystems.

### Anything you want to highlight for special attention from reviewers?

- This PR includes changes across all ecosystems. For a smoother review process, please refer to my commit messages as they provide clear information. Request to look the regex carefully for any missing manifest files which needs to be added to the `updated_files_regex`.

- Regex support for vendor files has been added only for the `go_modules` and `bundler` ecosystems, as these are the ones currently supported. [[doc](https://docs.github.com/en/code-security/dependabot/dependabot-version-updates/configuration-options-for-the-dependabot.yml-file#vendor)]

- Q. How changes to `updated_files_regex` will work ?
Before creating the PR, the `updated_files_regex` will validate that the files being updated are appropriate for the ecosystem. If any files are invalid, the PR will not be created and an error will be raised.

### How will you know you've accomplished your goal?

This PR should not cause any PR creation failures, except in cases where files from different ecosystems are being updated simultaneously. In those situations, an error will be raised to ensure that the updates are handled correctly.

### Checklist

- [x] I have run the complete test suite to ensure all tests and linters pass.
- [x] I have thoroughly tested my code changes to ensure they work as expected, including adding additional tests for new functionality.
- [x] I have written clear and descriptive commit messages.
- [x] I have provided a detailed description of the changes in the pull request, including the problem it addresses, how it fixes the problem, and any relevant details about the implementation.
- [x] I have ensured that the code is well-documented and easy to understand.

### CODE REVIEW

**Review by kbukum1 — APPROVED** [review]

**Comment by abdulapopoola** [comment]
@amazimbe , FYI I think this is related to the the issue you noticed today.

**Comment by abdulapopoola** [comment]
@honeyankit , is there a way to add tests so we can catch this earlier going forward?

**Comment by honeyankit** [comment]
> @honeyankit , is there a way to add tests so we can catch this earlier going forward?

@abdulapopoola This PR needs more work as `updated_files_regex` is not used anywhere in the Core as regex for every ecosystem is not up to update. I will add test that going forward adding `updated_files_regex` is compulsory else test will fail. This way we can keep track of the `updated_files_regex`

**Comment by abdulapopoola** [comment]
Thanks for this @honeyankit , a few generic questions:
1. Does this impact multi-dir in any way? Or grouping scenarios?
2. What is our testing / rollout plan to catch any regressions?

**Review by kbukum1 — APPROVED** [review]
Great job.

**Review by pavera — APPROVED** [review]

**Comment by honeyankit** [comment]
> Thanks for this @honeyankit , a few generic questions:
> 
> 1. Does this impact multi-dir in any way? Or grouping scenarios?
> 2. What is our testing / rollout plan to catch any regressions?

@abdulapopoola : These changes will be used in api which are behind the feature flag and will be a slow rollout to catch any regression.

File: nuget/spec/dependabot/nuget/file_updater_spec.rb [hunk 1]
```diff
@@ -110,6 +110,54 @@ def intercept_native_tools(discovery_content_hash:)
...
+          "Proj1/Proj1/Proj1.csproj"
+        ]
+
+        matching_files.each do |file_name|
+          expect(updated_files_regex).to(be_any { |regex| file_name.match?(regex) })
+        end
+      end
+
+      it "returns false for files that should not be updated" do
+        non_matching_files = [
```
Comment (brettfo) [inline nuget/spec/dependabot/nuget/file_updater_spec.rb:143]: Add `NuGet.Config` here as well as the lowercase version `nuget.config`.  Those files are important to NuGet updates, but we should never modify them in a PR.

**Review by brettfo — CHANGES_REQUESTED** [review]

File: nuget/spec/dependabot/nuget/file_updater_spec.rb (hunk 1 above)
Comment (honeyankit) [reply nuget/spec/dependabot/nuget/file_updater_spec.rb:143]: @brettfo : The ☝️ two files are added to the `non_matching_files` list.

**Review by brettfo — APPROVED** [review]

**Review by jakecoffman — APPROVED** [review]

**Pull request merged**